### PR TITLE
[TTAHUB-1648] internal goal duplicate cleanup

### PR DESCRIPTION
**Pull request by GarrettEHill**
First pass, uses oldest goal/objective per grant, not newest per grant or newest across grants.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1648

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/migrations/20230710000000-internal-goal-duplicate-cleanup.js
Comment (thewatermethod) [inline src/migrations/20230710000000-internal-goal-duplicate-cleanup.js:221]: won't we lose differences here? are we going with the oldest?

**Review by thewatermethod — COMMENTED** [review]
after running this with on newish prod snapshot

| table        | before | after | diff |
|--------------|--------|-------|------|
| goals        | 36680  | 35712 | -968 |
| ARG          | 45565  | 45565 | 0    |
| objectives   | 54049  | 53121 | -928 |
| arobjectives | 58480  | 58443 | -37  |

Do these numbers make sense? We are removing about 1000 (of 35000 & 54000 goals and objectives, respectively). This does not feel significant. 

Does it make sense that no AR Goals were removed?

**Comment by thewatermethod** [comment]
Script ran fine and it looks like everything makes sense on a read-through, although I'm really not qualified to evaluate the SQL. 

- Would look to get another review before merging.
- I wish there was something we could do to validate the removal gets the right stuff beyond the SQL the test itself runs.

**Review by AdamAdHocTeam — COMMENTED** [review]
@GarrettEHill let's go over this together when you are back in tomorrow.

**Review by kcirtapfromspace — APPROVED** [review]
+1 Great work @GarrettEHill 🎉

**Review by thewatermethod — APPROVED** [review]

File: src/migrations/20230802000000-internal-goal-duplicate-cleanup.js
Comment (hardwarehuman) [inline src/migrations/20230802000000-internal-goal-duplicate-cleanup.js:81]: Super unnecessary to do this on a dataset this small, but you can save expensive REGEX operations by doing
```
WITH unique_goal_names AS (
  SELECT DISTINCT name FROM "Goals"
),
smashed_names AS (
SELECT
  name,
  [regexes] "smashedName"
FROM unique_goal_names
)
SELECT
  g.id "goalId",
  "smashedName"
FROM "Goals" g
JOIN smashed_names sn
  ON g.name = sn.name
;
```

File: src/migrations/20230802000000-internal-goal-duplicate-cleanup.js
Comment (hardwarehuman) [inline src/migrations/20230802000000-internal-goal-duplicate-cleanup.js:2428]: This rolls back the transaction, correct? If not, there's also a way to make a sql error occur that rolls back the transaction if there are any records in `cascade`.

**Review by hardwarehuman — APPROVED** [review]

**Pull request merged**
